### PR TITLE
Support Zarr-Python 3

### LINKED ISSUE #1068: [titiler.xarray] update for Zarr 3.0
https://zarr.dev/blog/zarr-python-3-release/

we also need to pin zarr to 2.0 https://github.com/developmentseed/titiler/actions/runs/12695652933/job/35388000642?pr=1067

### PR DESCRIPTION
**Pull request by maxrjones**
Adds support for both opening Zarr stores using either Zarr-Python 2 or Zarr-Python 3.

Fixed #1068

### CODE REVIEW

File: src/titiler/xarray/pyproject.toml [hunk 1]
```diff
@@ -64,7 +64,16 @@ test = [
...
+]
+test-zarr-v2 = [
+    "pytest",
+    "pytest-cov",
+    "pytest-asyncio",
+    "httpx",
+    "zarr<3.0",
     "h5netcdf",
     "fsspec",
 ]
```
Comment (maxrjones) [inline src/titiler/xarray/pyproject.toml:79]: This is just for testing. I'm more familiar with hatch than pdm, but if we want more comprehensive tests for both minimum versions and/or upstream main branches I could work on setting that up.

File: src/titiler/xarray/titiler/xarray/io.py [hunk 2]
```diff
@@ -66,55 +52,8 @@ def xarray_open_dataset(  # noqa: C901
...
-
-        filesystem = fsspec.filesystem(protocol)  # type: ignore
-        file_handler = (
-            filesystem.open(src_path)
-            if xr_engine == "h5netcdf"
-            else filesystem.get_mapper(src_path)
-        )
-
-    else:
-        raise ValueError(f"Unsupported protocol: {protocol}, for {src_path}")
```
Comment (maxrjones) [inline src/titiler/xarray/titiler/xarray/io.py:117]: It's a lot simpler to rely on fsspec's import error message and use fsspec.filesystem(protocol). This also provides the benefit of supporting a wider range of protocols. Still, let me know if you'd rather control the error message through the removed approach.

File: src/titiler/xarray/titiler/xarray/io.py [hunk 3]
```diff
@@ -135,13 +74,21 @@ def xarray_open_dataset(  # noqa: C901
...
+        if _zarr_v3:
+            if protocol == "file":
+                store = zarr.storage.LocalStore(parsed.path, read_only=True)
+            else:
+                fs = fsspec.filesystem(protocol, storage_options={"asynchronous": True})
+                store = zarr.storage.FsspecStore(fs, path=src_path, read_only=True)
+            ds = xarray.open_zarr(store, **xr_open_args)
+        else:
+            fs = fsspec.filesystem(protocol)
+            ds = xarray.open_zarr(fs.get_mapper(src_path), **xr_open_args)
```
Comment (maxrjones) [inline src/titiler/xarray/titiler/xarray/io.py:91]: This is the bulk of the changes to account for Zarr no longer accepting fsspec's mutable mapping wrappers and expecting asynchronous file systems.

File: src/titiler/xarray/titiler/xarray/io.py (hunk 2 above)
Comment (vincentsarago) [reply src/titiler/xarray/titiler/xarray/io.py:117]: I think @abarciauskas-bgse wrote most of this originally in titiler-xarray

File: src/titiler/xarray/pyproject.toml [hunk 4]
```diff
@@ -64,7 +64,16 @@ test = [
...
+    "zarr",
+    "h5netcdf",
+    "fsspec",
+]
+test-zarr-v2 = [
+    "pytest",
+    "pytest-cov",
+    "pytest-asyncio",
+    "httpx",
+    "zarr<3.0",
```
Comment (vincentsarago) [inline src/titiler/xarray/pyproject.toml:76]: I would be fine only testing with zarr>3.0 TBH

File: src/titiler/xarray/titiler/xarray/io.py [hunk 5]
```diff
@@ -66,55 +52,8 @@ def xarray_open_dataset(  # noqa: C901
...
-        filesystem = fsspec.filesystem(protocol)  # type: ignore
-        file_handler = (
-            filesystem.open(src_path)
-            if xr_engine == "h5netcdf"
-            else filesystem.get_mapper(src_path)
-        )
-
-    else:
-        raise ValueError(f"Unsupported protocol: {protocol}, for {src_path}")
+        _zarr_v3 = module_available("zarr", minversion="3.0")
```
Comment (vincentsarago) [inline src/titiler/xarray/titiler/xarray/io.py:56]: can we move this a L82?

File: src/titiler/xarray/titiler/xarray/io.py (hunk 3 above)
Comment (vincentsarago) [inline src/titiler/xarray/titiler/xarray/io.py:91]:
```suggestion
        if module_available("zarr", minversion="3.0"):
            if protocol == "file":
                store = zarr.storage.LocalStore(parsed.path, read_only=True)
            else:
                fs = fsspec.filesystem(protocol, storage_options={"asynchronous": True})
                store = zarr.storage.FsspecStore(fs, path=src_path, read_only=True)
            
        else:
            store = fsspec.filesystem(protocol).get_mapper(src_path)

        ds = xarray.open_zarr(store, **xr_open_args)
```

File: src/titiler/xarray/pyproject.toml [hunk 6]
```diff
@@ -38,15 +38,15 @@ dependencies = [
...
+    "zarr",
     "h5netcdf",
     "fsspec",
     "s3fs",
     "aiohttp",
     "gcsfs",
 ]
 minimal = [
-    "zarr>=2,<3",
+    "zarr",
```
Comment (abarciauskas-bgse) [inline src/titiler/xarray/pyproject.toml:49]: will this work for zarr<3 ? Should that be specified?

File: src/titiler/xarray/titiler/xarray/io.py (hunk 2 above)
Comment (abarciauskas-bgse) [reply src/titiler/xarray/titiler/xarray/io.py:117]: This approach seems better to me too @maxrjones  I didn't know about fsspec's import error message and `fsspec.filesystem(protocol)`

File: src/titiler/xarray/pyproject.toml (hunk 6 above)
Comment (maxrjones) [reply src/titiler/xarray/pyproject.toml:49]: Yes, I accounted for both the new and old syntax in https://github.com/developmentseed/titiler/blob/e081d0f1452e4a950d8e9c13c939f0c5627956aa/src/titiler/xarray/titiler/xarray/io.py#L81-L89 and tested locally for zarr<3.
